React/React hooks.md

```diff
@@ -0,0 +1,386 @@
+  const memoizedValue = useMemo(() => someExpensiveCalculation(propA, propB), [propA, propB]);
+
+  // ...
+}
+```
+
+#### Wrap non-primitive values with useMemo, if you are using them as depenencies or sending them as props
+
+Since objects will be recreated on each render, they will also have a different reference every time - and `useEffect` will treat it as a changed dependency.
+
+```jsx
+export const MyComponent: FC = ({ propA, propB }) => {
+  const nonMemoizedObject = {
+    a: propA,
+    b: propB,
+  };
+
+  useEffect(() => {
+    /**
+      On each render, `useEffect` is causing an additional rerender,
+      since it's dependency is an object with a new reference every time.
+    */
+  }, [nonMemoizedObject])
+
+  // ...
+}
+```
+
+Assuming that `propA` and `propB` are primitive (or memoized, non-primitive) values, with `useMemo` you are ensuring that non-primitive references are retained throughout rerenders, until `propA` or `propB` have changed.
+
+```jsx
+export const MyComponent: FC = ({ propA, propB }) => {
+  const memoizedObject = useMemo(() => {
+    a: propA,
+    b: propB,
+  }, [propA, propB]);
+
+  useEffect(() => {
+    // Triggers rerenders only if object values have changed and `memoizedObject` is recreated
+  }, [memoizedObject])
+ 
+  // ...
+}
+```
+
+You can apply the same way of thinking if you need to pass an non-primitive value as a prop to a component and you know it will cause a large subtree to rerender each time it changes:
+
+```jsx
+export const MyComponent: FC = ({ propA, propB }) => {
+  const memoizedObject = useMemo(() => {
+    a: propA,
+    b: propB,
+  }, [propA, propB]);
+
+  const memoizedArray = useMemo(() => [propA, propB], [propA, propB])
+
+  return (
+    <ChildComponent objectProp={memoizedObject} arrayProp={memoizedArray}/>
+  );
+}
+```
+
+#### Do not use `useMemo` as a garantee that it will be a constant throughout component rerenders
+
+If you need a value to stay the same throughout rerenders, you might think of `useMemo` as a nice way to "mimick" a constant. While it might seem that way, [it is not garanteed](https://reactjs.org/docs/hooks-reference.html#usememo) that it will not get garbage collected and that the value will not be recalculated.
+
+```jsx
+export const MyComponent: FC = () => {
+  const id = useMemo(() => generateId(), []); // Not garanteed to be a constant
+
+  // ...
+}
+```
+
+Alternative approach, with lazy initialization:
+
+```jsx
+export const MyComponent: FC = () => {
+  const [id, _] = useState(() => generateId()); // Initial value will stay the same throughout rerenders
+
+  // ...
+}
+```
+
+You can also store the value in a ref:
+_Note: updating ref values will not trigger a rerender._
+
+```jsx
+export const MyComponent: FC = () => {
+  const id = useRef(null);
+
+  if (!id.current) {
+    id.current = generateId();
+  }
+
+  // ...
+}
+```
+
+If you want the component to rerender when the value is set:
```

Comment:
You should never want that.
This will trigger two renders.
Remember the React hooks flow.
https://github.com/donavon/hook-flow
`useEffect` is executed at the end, when the first paint is already done, so this will trigger the second render, which is bad for performance.
This is a BAD example and we should note this in the handbook.
Don't ever do this!

Reply:
That reminds me that we should add a chapter about hooks flow and include this chart

Reply:
This is fixed in #109